crossbeam-queue/src/spsc.rs

```diff
@@ -0,0 +1,449 @@
+use std::cell::Cell;
+use std::fmt;
+use std::marker::PhantomData;
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use crossbeam_utils::CachePadded;
+
+use err::{PopError, PushError};
+
+/// The inner representation of a single-producer single-consumer queue.
+struct Inner<T> {
+    /// The head of the queue.
+    ///
+    /// This integer is in range `0 .. 2 * cap`.
+    head: CachePadded<AtomicUsize>,
+
+    /// The tail of the queue.
+    ///
+    /// This integer is in range `0 .. 2 * cap`.
+    tail: CachePadded<AtomicUsize>,
+
+    /// The buffer holding slots.
+    buffer: *mut T,
+
+    /// The queue capacity.
+    cap: usize,
+
+    /// Indicates that dropping a `Buffer<T>` may drop elements of type `T`.
+    _marker: PhantomData<T>,
+}
+
+impl<T> Inner<T> {
+    /// Returns a pointer to the slot at position `pos`.
+    ///
+    /// The position must be in range `0 .. 2 * cap`.
+    #[inline]
+    unsafe fn slot(&self, pos: usize) -> *mut T {
+        if pos < self.cap {
+            self.buffer.add(pos)
+        } else {
+            self.buffer.add(pos - self.cap)
+        }
+    }
+
+    /// Increments a position by going one slot forward.
+    ///
+    /// The position must be in range `0 .. 2 * cap`.
+    #[inline]
+    fn increment(&self, pos: usize) -> usize {
+        if pos < 2 * self.cap - 1 {
+            pos + 1
+        } else {
+            0
+        }
+    }
+
+    /// Returns the distance between two positions.
+    ///
+    /// Positions must be in range `0 .. 2 * cap`.
+    #[inline]
+    fn distance(&self, a: usize, b: usize) -> usize {
+        if a <= b {
+            b - a
+        } else {
+            2 * self.cap - a + b
+        }
+    }
+}
+
+impl<T> Drop for Inner<T> {
+    fn drop(&mut self) {
+        let mut head = self.head.load(Ordering::Relaxed);
+        let tail = self.tail.load(Ordering::Relaxed);
+
+        // Loop over all slots that hold a value and drop them.
+        while head != tail {
+            unsafe {
+                self.slot(head).drop_in_place();
+            }
+            head = self.increment(head);
+        }
+
+        // Finally, deallocate the buffer, but don't run any destructors.
+        unsafe {
+            Vec::from_raw_parts(self.buffer, 0, self.cap);
+        }
+    }
+}
+
+/// Creates a bounded single-producer single-consumer queue with the given capacity.
+///
+/// Returns the producer and the consumer side for the queue.
+///
+/// # Panics
+///
+/// Panics if the capacity is zero.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_queue::spsc;
+///
+/// let (p, c) = spsc::<i32>(100);
+/// ```
+pub fn spsc<T>(cap: usize) -> (Producer<T>, Consumer<T>) {
```

Comment:
It looks like the input argument really is a `NonZeroUsize`.

Reply:
True, but `NonZeroUsize` would be more appropriate for memory layout optimizations rather than ad-hoc uses like this one. There's little advantage of `spsc(NonZeroUsize::new(n).unwrap())` over `spsc(n)` :)

Reply:
For completeness, the rationale for my comment above was to remove the runtime-panic behavior for something that can be encoded at type-level. I'm fine with your call if you think it is not worth it for ergonomics.